app/src/main/java/com/keylesspalace/tusky/components/notifications/NotificationsViewModel.kt

```diff
@@ -0,0 +1,402 @@
+import androidx.room.withTransaction
+import at.connyduck.calladapter.networkresult.NetworkResult
+import at.connyduck.calladapter.networkresult.fold
+import at.connyduck.calladapter.networkresult.map
+import at.connyduck.calladapter.networkresult.onFailure
+import com.keylesspalace.tusky.appstore.EventHub
+import com.keylesspalace.tusky.appstore.PreferenceChangedEvent
+import com.keylesspalace.tusky.components.preference.PreferencesFragment.ReadingOrder
+import com.keylesspalace.tusky.components.timeline.Placeholder
+import com.keylesspalace.tusky.components.timeline.toEntity
+import com.keylesspalace.tusky.components.timeline.util.ifExpected
+import com.keylesspalace.tusky.components.timeline.viewmodel.TimelineViewModel
+import com.keylesspalace.tusky.db.AccountManager
+import com.keylesspalace.tusky.db.AppDatabase
+import com.keylesspalace.tusky.entity.Filter
+import com.keylesspalace.tusky.entity.Notification
+import com.keylesspalace.tusky.network.FilterModel
+import com.keylesspalace.tusky.network.MastodonApi
+import com.keylesspalace.tusky.settings.PrefKeys
+import com.keylesspalace.tusky.usecase.TimelineCases
+import com.keylesspalace.tusky.util.EmptyPagingSource
+import com.keylesspalace.tusky.util.deserialize
+import com.keylesspalace.tusky.util.serialize
+import com.keylesspalace.tusky.viewdata.NotificationViewData
+import com.keylesspalace.tusky.viewdata.StatusViewData
+import com.keylesspalace.tusky.viewdata.TranslationViewData
+import javax.inject.Inject
+import kotlinx.coroutines.Deferred
+import kotlinx.coroutines.Dispatchers
+import kotlinx.coroutines.Job
+import kotlinx.coroutines.asExecutor
+import kotlinx.coroutines.async
+import kotlinx.coroutines.flow.MutableStateFlow
+import kotlinx.coroutines.flow.StateFlow
+import kotlinx.coroutines.flow.asStateFlow
+import kotlinx.coroutines.flow.combine
+import kotlinx.coroutines.flow.flowOn
+import kotlinx.coroutines.launch
+import retrofit2.HttpException
+
+class NotificationsViewModel @Inject constructor(
+    private val timelineCases: TimelineCases,
+    private val api: MastodonApi,
+    eventHub: EventHub,
+    private val accountManager: AccountManager,
+    private val preferences: SharedPreferences,
+    private val filterModel: FilterModel,
+    private val db: AppDatabase,
+) : ViewModel() {
+
+    private val _filters = MutableStateFlow(
+        accountManager.activeAccount?.let { account -> deserialize(account.notificationsFilter) } ?: emptySet()
+    )
+    val filters: StateFlow<Set<Notification.Type>> = _filters.asStateFlow()
+
+    /** Map from notification id to translation. */
+    private val translations = MutableStateFlow(mapOf<String, TranslationViewData>())
+
+    private var remoteMediator = NotificationsRemoteMediator(accountManager, api, db, filters.value)
+
+    private var readingOrder: ReadingOrder =
+        ReadingOrder.from(preferences.getString(PrefKeys.READING_ORDER, null))
+
+    @OptIn(ExperimentalPagingApi::class)
+    val notifications = Pager(
```

Comment:
If you still want to refresh the paging data from the ViewModel, here's a way to do it:

```
    private val refreshTrigger = MutableStateFlow(0L)

    val notifications: Flow<...> = refreshTrigger.flatMapLatest {
        Pager(
            config = PagingConfig(pageSize = LOAD_AT_ONCE),
            remoteMediator = remoteMediator,
            pagingSourceFactory = { ... }
        ).flow
      }.cachedIn(viewModelScope)

    private fun refresh() {
        refreshTrigger.value++
    }
```

Reply:
thats what I was looking for, thank you!